### PR TITLE
feat(editor): improve documentation editor using Remirror

### PR DESCRIPTION
**Pull request by ngamanda**
## Highlights
editor-demo
- Improved UX for documentation editing with 
  - Modals to Add Links / Images
  - Inline menus for quicker access to style text
  - Inline link previews
- @mentions: users can now tag entities by typing `@`. More details about the entity shown upon hovering over the mention
- Resizable images: users can now resize images
  - By default, uploading images will be saved as markdown (i.e. `alt text`). Upon resizing, the image will be stored as HTML Markup (i.e. `alt text`)
- Table editing
  - Controls to add/remove table rows/columns easily
  - Tables are converted to markdown where possible. Otherwise, it will be stored as in HTML Markup. (Markdown tables cannot contain lists, headings, code blocks, or another table)

## Summary of Changes
- Install additional packages - `@remirror/*` for editor, `prosemirror-autocomplete` for powering inline suggestions, `turndown-plugin-gfm` to add `gfm` support for tables
- Add custom buttons/controls to fit with AntD stylings
- Add new Remirror DataHubMentionsExtension to contain rendering logic and to import `prosemirror-autocomplete` plugin
- Remove sanitising of documentations located outside the editor (already handled by Remirror)
- Update Jest configuration to explicitly transpile `lib0` and `y-protocols`, keeping `'^.+\\.module\\.(css|sass|scss)$'` from default configurations

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/Editor.tsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    return (
+        <EditorContainer>
+            <ThemeProvider theme={EditorTheme}>
+                <Remirror classNames={['ant-typography']} editable={!readonly} manager={manager} initialContent={state}>
+                    {!readonly && (
+                        <>
+                            <Toolbar />
+                            <CodeBlockToolbar />
+                            <FloatingToolbar />
+                            <TableComponents tableCellMenuProps={{ Component: TableCellMenu }} />
```
Comment (chriscollins3456) [inline datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/Editor.tsx:87]: is it possible to resize columns and rows of the table with your mouse? I'm trying to do it locally but not figuring out how if it is possible

**Review by chriscollins3456 — COMMENTED** [review]
This is looking incredible! I have some comments here and there, some suggestions and some requests for things like docs and tests. However overall this is looking so good.

I pulled it down myself and am having a blast with it locally! You guys absolutely killed this. After you get back to us with my comments / suggestions let's work to get this in soon!

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/Editor.tsx (hunk 1 above)
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/Editor.tsx:87]: Actually it's possible and it's enabled in Grab's fork. The downside is that the table will be stored in HTML rather than Markdown format. I wasn't sure if DataHub wanted that activated since you can't 'un-resize' a table (or we could also introduce a button to 'un-resize' it but I'll need more time to investigate it).

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/Editor.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            ...(readonly ? [] : [new HistoryExtension()]),
+        ],
+        content,
+        stringHandler: 'markdown',
+    });
+
+    useImperativeHandle(ref, () => getContext(), [getContext]);
+    useMount(() => {
+        manager.view.focus();
+    });
```
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/Editor.tsx:76]: Haha yup, [react-use](https://github.com/streamich/react-use) is a lifesaver!

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/Editor.tsx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import { CodeBlockToolbar } from './toolbar/CodeBlockToolbar';
+import { FloatingToolbar } from './toolbar/FloatingToolbar';
+import { Toolbar } from './toolbar/Toolbar';
+import { OnChangeMarkdown } from './OnChangeMarkdown';
+import { TableCellMenu } from './toolbar/TableCellMenu';
+import { DataHubMentionsExtension } from './extensions/mentions/DataHubMentionsExtension';
+import { MentionsComponent } from './extensions/mentions/MentionsComponent';
+
+type EditorProps = {
+    readonly?: boolean;
```
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/Editor.tsx:38]: Will make the change

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/EditorTheme.tsx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    ${extensionCalloutStyledCss}
+    ${extensionCodeBlockStyledCss}
+    ${extensionCountStyledCss}
+    ${extensionGapCursorStyledCss}
+    ${extensionImageStyledCss}
+    ${extensionListStyledCss}
+    ${extensionMentionAtomStyledCss}
+    ${extensionPlaceholderStyledCss}
+    ${extensionPositionerStyledCss}
+    ${extensionTablesStyledCss}
```
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/EditorTheme.tsx:61]: We'll need to import the css blocks from Remirror _only if_ we want their default styles. We can choose to not import it if we want to style it ourselves.

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/OnChangeMarkdown.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    onChange: (md: string) => void;
+}
+
+export const OnChangeMarkdown = ({ onChange }: OnChangeMarkdownProps): null => {
+    const { getMarkdown } = useHelpers();
+
+    const onDocChanged = useCallback(
+        ({ state }) => {
+            const markdown = getMarkdown(state);
+            onChange(markdown);
```
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/OnChangeMarkdown.tsx:15]: Yup, internally Remirror stores it as JSON/HTML (They provide 2 output formats). 

The [MarkdownExtension](https://remirror.io/docs/api/extension-markdown#property-htmltomarkdown) provides two configurable functions `htmlToMarkdown` (used to convert Remirror to Markdown) and `markdownToHtml` (used to convert passed-in Markdown to Remirror). Remirror/We are using libraries for this parsing.

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/extensions/htmlToMarkdown.tsx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            });
+            const urn = node.getAttribute(DATAHUB_MENTION_ATTRS.urn);
+            if (!urn) return '';
+
+            return `[${node.textContent}](${urn})`;
+        },
+    });
+
+/* Converts the provide HTML to markdown. */
+export function htmlToMarkdown(html: string): string {
```
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/extensions/htmlToMarkdown.tsx:98]: Sure! Will add some unit test.

Theorectially, this file would only be updated if we're building more customisations on top of the [GFM spec](https://github.github.com/gfm/) since the [turndown library](https://www.npmjs.com/package/turndown) used already assists in parsing most of the content from HTML to Markdown. 

Additional rules are being added at the moment to force turndown to skip parsing some content (i.e. resizable images, lists in tables, etc)

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/extensions/mentions/MentionsDropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        return [...acc, ...options];
+    }, []);
+};
+
+export const MentionsDropdown = ({ suggestions: _suggestions }: Props) => {
+    const entityRegistry = useEntityRegistry();
+    const [options, setOptions] = useState<Option[]>([]);
+    const { createDataHubMention } = useCommands();
+
+    useDebounce(() => setOptions(flattenOptions(_suggestions)), 250, [_suggestions]);
```
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/extensions/mentions/MentionsDropdown.tsx:67]: Ah I did a refactor and missed out renaming this back, will make the change. Thanks for catching it!

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/toolbar/FloatingToolbar.tsx [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+const LinkText = styled(Text)`
+    padding-left: 4px;
+    max-width: 250px;
+    text-overflow: ellipsis;
+    white-space: nowrap;
+    overflow: hidden;
+`;
+
+export const FloatingToolbar = () => {
```
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/toolbar/FloatingToolbar.tsx:41]: Will fix this!

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/toolbar/Toolbar.tsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                onClick={() => commands.toggleCodeBlock()}
+            />
+            <Divider type="vertical" />
+            <AddImageButton />
+            <AddLinkButton />
+            <CommandButton
+                icon={<TableOutlined />}
+                commandName="createTable"
+                onClick={() => commands.createTable()}
+                disabled={active.table()} /* Disables nested tables */
```
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/toolbar/Toolbar.tsx:95]: By default Markdown doesn't support nested tables. And I'm not sure about the capabilities for other Markdown viewers to render nested HTML tables (should any other DataHub users choose to use their own viewers). IMO the usage of nested tables is really low so I kept it out to reduce complexity.

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/DocumentationTab.tsx [hunk 10]
```diff
@@ -16,24 +14,24 @@ import { LinkList } from './components/LinkList';
...
     hideLinksButton?: boolean;
 }
 
 export const DocumentationTab = ({ properties }: { properties?: Props }) => {
+    const editorRef = useRef<any>();
     const hideLinksButton = properties?.hideLinksButton;
     const { urn, entityData } = useEntityData();
     const refetch = useRefetch();
     const description = entityData?.editableProperties?.description || entityData?.properties?.description || '';
-    const sanitizedDescription = DOMPurify.sanitize(description);
```
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/DocumentationTab.tsx:36]: Hmm, I remembered the double sanitisation did throw out some unwanted syntax in the output. Let me take a look at it again.

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/DescriptionEditor.tsx [hunk 11]
```diff
@@ -140,12 +143,9 @@ export const DescriptionEditor = ({ onComplete }: { onComplete?: () => void }) =
...
                 </Button>
             </TabToolbar>
-            <StyledMDEditor
-                value={updatedDescription}
-                onChange={(v) => handleEditorChange(v || '')}
-                preview="live"
-                visiableDragbar={false}
-            />
+            <EditorContainer>
+                <Editor content={updatedDescription} onChange={(v) => handleEditorChange(v)} />
```
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/DescriptionEditor.tsx:147]: Yup we could! But not sure if the editor would render well in a modal. Perhaps in the next iteration?

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/DocumentationTab.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/DocumentationTab.tsx:36]: okay if that's the case I don't think it's a problem as long as we feel good about remirror being able to handle this for us

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/toolbar/Toolbar.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/toolbar/Toolbar.tsx:95]: makes sense to me! thanks for the explanation

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/DescriptionEditor.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/DescriptionEditor.tsx:147]: would you mind if we try and see how it looks in the modal? hopefully we can make some adjustments if it's not looking great, but I just think it would be confusing to have two different rich text editors in the app. Especially when one is much nicer than the other - people may be disappointed when trying to edit schema field documentation from the modal for example.

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/extensions/htmlToMarkdown.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/extensions/htmlToMarkdown.tsx:114]: okay cool beans sounds good!

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/DescriptionEditor.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/DescriptionEditor.tsx:147]: Hey @chriscollins3456, I've pushed a new commit that reflects the new editor across the app, looks good so far. Let me know what you think!

[the remark this answers is not in the document]
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/DescriptionEditor.tsx:147]: amazing! thanks so much :)

**Review by chriscollins3456 — APPROVED** [review]
This is looking so great!

Thanks so much for getting back to my comments and adding all the tests and what not. I pulled down again and tried locally and things are looking awesome :)

I have one more comment about making links `target="_blank"` so they open in a new tab, but otherwise I think this thing is ready to rumble!

I did notice that there's a few conflicts with master, though, so once those are resolved and we get CI to pass everything, let's get this guy merged.

File: datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/toolbar/LinkModal.tsx [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+import { FromToProps } from '@remirror/core-types';
+import { useAttrs, useCommands, useEditorState, useHelpers } from '@remirror/react';
+import { getMarkRange } from '@remirror/core-utils';
+
+type LinkModalProps = {
+    visible: boolean;
+    handleClose: () => void;
+};
+
+export const LinkModal = (props: LinkModalProps) => {
```
[the remark this answers is not in the document]
Comment (ngamanda) [reply datahub-web-react/src/app/entity/shared/tabs/Documentation/components/editor/toolbar/LinkModal.tsx:12]: Sure, fixed in https://github.com/datahub-project/datahub/pull/6631/commits/bd3d8d25891e8098aeaa4ea1c4670070c5cbfaa8

File: .github/workflows/docker-unified.yml [hunk 13]
```diff
@@ -314,6 +314,8 @@ jobs:
         run: |
           ./gradlew :datahub-frontend:dist -x test -x yarnTest -x yarnLint --parallel
           mv ./datahub-frontend/build/distributions/datahub-frontend-*.zip datahub-frontend.zip
+        env:
+          NODE_OPTIONS: "--max-old-space-size=3072"
```
Comment (chriscollins3456) [inline .github/workflows/docker-unified.yml:318]: why is this change necessary?

Comment (ngamanda) [reply .github/workflows/docker-unified.yml:318]: Oh it's to resolve the failing build that was caused by `FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory.` in https://github.com/datahub-project/datahub/actions/runs/3804191769/jobs/6471221050. This line simply increases the memory limit allocated to node. If there are other solutions please feel free to make changes to the PR too!

Comment (chriscollins3456) [reply .github/workflows/docker-unified.yml:318]: it looks like this failure is still happening :( do you know if we need to put this type of change elsewhere?

Comment (chriscollins3456) [reply .github/workflows/docker-unified.yml:318]: another initiative we've discussed is removing the unneeded dependencies from our package.json - unsure how many things we can remove but if we can shrink our dependencies this may help solve this problem! it's something i'm looking into and will see if it helps here

Comment (ngamanda) [reply .github/workflows/docker-unified.yml:318]: Hmm, the latest build passed though: https://github.com/datahub-project/datahub/actions/runs/3805789729/jobs/6559405831. The only pipeline that's failing is https://github.com/datahub-project/datahub/actions/runs/3805789729/jobs/6559407548 which looks unrelated to the documentation, probably a flaky test (but I can't rereun it :x)

Comment (ngamanda) [reply .github/workflows/docker-unified.yml:318]: Also let me know if you need help looking into shrinking the dependencies 🗡️ !

Comment (chriscollins3456) [reply .github/workflows/docker-unified.yml:318]: ah so it looks like it's fixed in the Docker, Build, Scan, Test runs, but i'm still seeing the error in the logs on the general build & test / build job (https://github.com/datahub-project/datahub/actions/runs/3805789734/jobs/6521531814) - you'll have to search for it cuz that's a huge log lol

also it seems that that's timing out after 60 minutes, my hunch is that it's related to this build failing or even that something is taking much longer for some reason.

**Comment by chriscollins3456** [comment]
everything is passing in CI now woo! except for these failing Monitoring tests but those are failing for all open PRs

**Comment by ngamanda** [comment]
Awesome, thanks for the help @chriscollins3456 !

**Review by jjoyce0510 — APPROVED** [review]
I finally got a chance to take a deep look at this PR.

All I have to say is: WOW. The new experience looks and feels incredible - A major improvement on usability that will benefit many people (thousands of active users) in the community directly!

This PR represents an incredible piece of collaboration between Grab and the DataHub Community. I want to say thank you to @ngamanda, @chriscollins3456, and of course Harvey on behalf of the entire DataHub Team and the broader community. I look forward to continued partnership between our teams as we continue to improve DataHub.

Very happy to give this a final "LGTM" stamp. 

Cheers!
John
